pkg/services/ngalert/accesscontrol/silences.go

```diff
@@ -97,157 +97,219 @@ func NewSilenceService(ac ac.AccessControl, store RuleUIDToNamespaceStore) *Sile
-	for _, silence := range silences {
-		ruleUID := silence.GetRuleUID()
-		if ruleUID == nil { // if this is a general silence
-			result = append(result, silence)
-			continue
-		}
-		key := *ruleUID
-		silencesByRuleUID[key] = append(silencesByRuleUID[key], silence)
+	if canAll {
+		return silences, nil
 	}
-	if len(silencesByRuleUID) == 0 { // if only general silences are provided no need in other checks
-		return result, nil
-	}
-	namespacesByRuleUID, err := s.store.GetNamespacesByRuleUID(ctx, user.GetOrgID(), maps.Keys(silencesByRuleUID)...)
+
+	silencesWithFolders, err := s.withFolders(ctx, user.GetOrgID(), silences...)
 	if err != nil {
 		return nil, err
 	}
 
+	result := make([]*models.Silence, 0, len(silences))
 	namespacesByAccess := make(map[string]bool) // caches results of permissions check for each namespace to avoid repeated checks for the same folder
-	for ruleUID, silence := range silencesByRuleUID {
-		ns, ok := namespacesByRuleUID[ruleUID]
-		if !ok { // this means that there is no rule with such UID.
-			continue
-		}
-		hasAccess, ok := namespacesByAccess[ns]
+	for _, silWithFolder := range silencesWithFolders {
+		hasAccess, ok := namespacesByAccess[silWithFolder.folderUID]
 		if !ok {
-			hasAccess, err = s.HasAccess(ctx, user, readRuleSilenceEvaluator(ns))
-			if err != nil {
-				return nil, err
+			hasAccess = s.authorizeReadSilence(ctx, user, silWithFolder) == nil
+
+			// Cache non-empty namespaces to avoid repeated checks for the same folder.
+			if silWithFolder.folderUID != "" {
+				namespacesByAccess[silWithFolder.folderUID] = hasAccess
 			}
-			namespacesByAccess[ns] = hasAccess
 		}
 		if hasAccess {
-			result = append(result, silence...)
+			result = append(result, silWithFolder.Silence)
 		}
 	}
 	return result, nil
 }
 
-// AuthorizeReadSilence checks if user has access to read a silence
+// AuthorizeReadSilence checks if user has access to read a silence.
 func (s SilenceService) AuthorizeReadSilence(ctx context.Context, user identity.Requester, silence *models.Silence) error {
-	canAll, err := s.HasAccess(ctx, user, readAllSilencesEvaluator)
+	canAll, err := s.authorizeReadSilencePreConditions(ctx, user)
 	if canAll || err != nil { // return early if user can either read all silences or there is error
 		return err
 	}
 
+	silWithFolder, err := s.withFolders(ctx, user.GetOrgID(), silence)
+	if err != nil || len(silWithFolder) != 1 {
+		return fmt.Errorf("resolve rule UID to folder UID: %w", err)
+	}
+
+	return s.authorizeReadSilence(ctx, user, silWithFolder[0])
+}
+
+// authorizeReadSilencePreConditions checks necessary preconditions for reading silences. Returns true if user can
+// read all silences. Returns error if user does not have access to read any silences.
+func (s SilenceService) authorizeReadSilencePreConditions(ctx context.Context, user identity.Requester) (bool, error) {
+	canAll, err := s.HasAccess(ctx, user, readAllSilencesEvaluator)
```

Comment:
I wonder if we can hack into user's permissions and check if user has wildcard scope, and set canAll to true if it has.

Reply:
Added silence read folder wildcard scope to the readAllSilencesEvaluator: https://github.com/grafana/grafana/pull/87998/commits/c214702f11b26fe788c4ae415e6ac74e570dfe41

This is not technically checking for all forms of wildcards, but should cover the main one without having to manually check user permissions.